### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by LiveMethod**
Significant, opinionated refactor, organizing the codebase, commenting, and changing things to better suit the way I plan to use this repo (less CLI, more direct args and logging)

The notebook has been deleted, since I'm not planning to use it, and it included lots of duplication of, but also lots of divergence from, the training file.

Total rewrite of the data visualization.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -162,4 +162,5 @@ cython_debug/
 #.idea/
 
 # System
-.DS_Store
\ No newline at end of file
+.DS_Store
+models/**
```
Comment (LiveMethod) [inline .gitignore:166]: this will fill up with generated files. did force inclusion of the example dir.

File: compare.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        sr_t, train = wavfile.read(self.train_wav)
+        sr_p, pred = wavfile.read(self.pred_wav)
+        if sr_d != sr_t or sr_t != sr_p:
+            raise ValueError("Sample rates of the wav files do not match. Are they from the same session?")
+        else:
+            self.sr = sr_p
+        
+        # File lengths can sometimes differ by a sample or two. This will even
+        # out a negligible difference but throw for anything noteworthy
+        shortest_data = min(len(dry), len(train), len(pred))
```
Comment (LiveMethod) [inline compare.py:61]: Bumped into this issue with the test data. Wrote around it slightly doesn't seem problematic.

File: compare.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        x_range = np.linspace(0, len(dry) / sr, num=len(dry))
+        divider = make_axes_locatable(ax1)
+        ax1_bottom = divider.append_axes("bottom", size="55%", pad=0.25)
+
+        # Custom divergent colormap from red to white to green
+        rwg = [(0.7, 0, 0), (1, 1, 1), (0, 0.6, 0)]
+        nodes = [0.0, 0.5, 1.0]
+        diff_cmap = LinearSegmentedColormap.from_list('rwg', list(zip(nodes, rwg)))
+        plt.colormaps.register(cmap=diff_cmap, name='rwg')
```
Comment (LiveMethod) [inline compare.py:98]: So matplotlib will give us `turbo` but not `rwg` ayfs lol

File: compare.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        # ==============================================================================
+        # Bottom right (AX4): Waveform comparison
+        # ==============================================================================
+        overlap_waveform = np.minimum(pred, train)
+
+        ax4.set_title('Waveform Comparison')
+        ax4.plot(x_range, pred, label='Pred +', color='green', alpha=0.5)
+        ax4.plot(x_range, train, label='Train +', color='red', alpha=0.5)
+        ax4.plot(x_range, overlap_waveform, label='both', color='white', alpha=1)
+        ax4.plot(x_range, dry, label='Dry', color='lightgray', alpha=1)
```
Comment (LiveMethod) [inline compare.py:193]: Hacky but kinda slick: maps both waveforms in diff colors but then maps their intersection in white to effectively erase everything that isn't the diff. This leaves room to map the dry signal in the same place.

File: compare.py
Comment (LiveMethod) [inline compare.py:1]: result

Number of new visualizations.
* Frequency graph from the previous plot setup, but also displayed with a log-x axis.
* Power spectrum density. This was an accident but turned out really illustrative. Shows a sharp diversion at the 5k mark, i wonder if that's the prefilter?
* Mel spectrogram diff, showing frequency areas that are more or less represented in each file. Exposes some interesting artifacting in the low end. I haven't graphed the source data directly, or one of the many-epoch builds, so RTBS whether this improves with additional training.
* A take on the original waveform graphic, described in a bit more detail below.

File: predict.py [hunk 5]
```diff
@@ -4,34 +4,14 @@
...
-    
-def error_to_signal(y_true, y_pred): 
-    y_true, y_pred = pre_emphasis_filter(y_true), pre_emphasis_filter(y_pred)
-    return K.sum(tf.pow(y_true - y_pred, 2), axis=0) / K.sum(tf.pow(y_true, 2), axis=0) + 1e-10
-
-def normalize(data):
-    data_max = max(data)
-    data_min = min(data)
-    data_norm = max(data_max,abs(data_min))
-    return data / data_norm
```
Comment (LiveMethod) [inline predict.py:34]: Available from the shared utils/data tools

File: predict.py [hunk 6]
```diff
@@ -4,34 +4,14 @@
...
 import h5py
 import argparse
 
-
-def save_wav(name, data):
-    if name.endswith('.wav') == False:
-        name = name + '.wav'
-    wavfile.write(name, 44100, data.flatten().astype(np.float32))
-    print("Predicted wav file generated: "+name)
-
```
Comment (LiveMethod) [inline predict.py:22]: Pulled in favor of just using the stock one directly

File: utils/data.py
Comment (LiveMethod) [inline utils/data.py:1]: Things like prefilters and ESR calcs were done in 3+ places, each slightly different, and expecting slightly different data types. This provides a single shared version of all of them that can accept both np and tf data.

File: utils/data.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        ---
+        ESR measures the error between predicted and actual signals.
+        https://www.mdpi.com/2076-3417/10/3/766/htm
+        """
+        if use_filter:
+            y, y_pred = Data.pre_emphasis_filter(y), Data.pre_emphasis_filter(y_pred)
+        if isinstance(y, np.ndarray):
+            return np.sum(np.power(y - y_pred, 2)) / (np.sum(np.power(y, 2) + 1e-10))
+        else:
+            return tf.reduce_sum(tf.square(y - y_pred)) / (tf.reduce_sum(tf.square(y)) + 1e-10)
```
Comment (LiveMethod) [inline utils/data.py:46]: paren/pemdas issue resolved that was present in the main file but fixed in the notebook, IIRC

File: train.py [hunk 8]
```diff
@@ -1,216 +1,228 @@
...
+
+def main(
+    source_dir:str,
+    training_mode:Literal['fast', 'accurate', 'extended'] = 'fast',
+    split_data:int = 1,
+    batch_size:int = 4096,
+    input_size:int = 100,
+    max_epochs:int = 1,
+    test_pct:int = 20
+):
```
Comment (LiveMethod) [inline train.py:30]: Takes more explicit args in favor of CLI args

File: train.py [hunk 9]
```diff
@@ -1,216 +1,228 @@
...
+        Speed training (default)
+        Accuracy training
+        Extended training (set max_epochs as desired, for example 50+)
+    """
+    print(f"Training started")
+
+    timestamp = datetime.datetime.now().strftime("%Y-%b-%d-%H%M%S").lower()
+    name, in_files, out_files = IO.validate_source(source_dir)
+    out_dir = f'{name}_{timestamp}'
+    os.makedirs(f'models/{out_dir}')
```
Comment (LiveMethod) [inline train.py:59]: Prev version would fail on name collision w folder, which made it difficult to re-run similar models back to back. This removes collision risk via timestamps in the folder names. Ugly but useful.

File: train.py [hunk 10]
```diff
@@ -1,216 +1,228 @@
...
         hidden_units= 96
-
-    # Create Sequential Model ###########################################
+    elif training_mode == 'accurate':
+        # ~10x longer than Speed Training
+        learning_rate = 0.01
+        conv1d_strides = 4
+        conv1d_filters = 36
+        hidden_units= 64
+    else:
```
Comment (LiveMethod) [inline train.py:74]: Switched around to match the args, so fast is the default case

File: train.py [hunk 11]
```diff
@@ -1,216 +1,228 @@
...
+    #
+    # The history object is used to store the results of model.fit and append
+    # metadata about the training process to the saved model file.
+    # =========================================================================
     clear_session()
+    history = None
     model = Sequential()
-    model.add(Conv1D(conv1d_filters, 12,strides=conv1d_strides, activation=None, padding='same',input_shape=(input_size,1)))
-    model.add(Conv1D(conv1d_filters, 12,strides=conv1d_strides, activation=None, padding='same'))
+    model.add(Input(shape=(input_size, 1)))
```
Comment (LiveMethod) [inline train.py:95]: There was a warning or something saying that this is the preferred way to set inputs, rather than in layer args. Works fine so sure.

File: train.py [hunk 12]
```diff
@@ -1,216 +1,228 @@
...
+    out_sr, out_data = wavfile.read(out_files[0])
+
+    # Validate that sample rates match
+    if(in_sr != out_sr):
+        # It's pretty trivial to do this conversion, but given that it's
+        # a strange situation to be in, preference on raising, as it
+        # probably indicates that something weird is going on.
+        raise ValueError("Input and output files have different sample rates.")
+    else:
+        sr = in_sr
```
Comment (LiveMethod) [inline train.py:126]: sample rates weren't being used much and there were hardcoded values, so this confirms that there's one known-good sample rate, and then uses that everywhere.

File: train.py [hunk 13]
```diff
@@ -1,216 +1,228 @@
...
-    parser.add_argument("--batch_size", type=int, default=4096)
-    parser.add_argument("--max_epochs", type=int, default=1)
-    parser.add_argument("--create_plots", type=int, default=1)
-    parser.add_argument("--input_size", type=int, default=100)
-    parser.add_argument("--split_data", type=int, default=1)
-    args = parser.parse_args()
-    main(args)
\ No newline at end of file
+    x_test_data = x_all[-int(len(x_all) * test_size):]
+    y_test_data = y_all[-int(len(y_all) * test_size):]
```
Comment (LiveMethod) [inline train.py:183]: Linked  to test_size instead of hardcoded, slightly prettified

File: train.py [hunk 14]
```diff
@@ -1,216 +1,228 @@
...
+    # append is not forward compatible. This suppresses the warning.
+    # Amenable to reviewing h5 metadata usage and dropping if unused or
+    # switching to a lib that does equivalent things for keras files.
+    # =========================================================================
+
+    model_save_path = f'models/{out_dir}/{name}.h5'
+
+    sys.stderr = open(os.devnull, "w")  # silence stderr
+    model.save(model_save_path)
+    sys.stderr = sys.__stderr__  # unsilence stderr
```
Comment (LiveMethod) [inline train.py:212]: There are less hamfisted ways to suppress warnings, but it looks like tf/keras overrides those (i know scipy does). This routes errors into the void for just the one call, then puts them back.

File: train.py [hunk 15]
```diff
@@ -1,216 +1,228 @@
...
+
+        grp.attrs['input_size'] = input_size
+        grp.attrs['epochs'] = max_epochs
+        grp.attrs['batch_size'] = batch_size
+        grp.attrs['learning_rate'] = learning_rate
+        grp.attrs['model_architecture'] = str(model.to_json())
+        grp.attrs['data_preprocessing'] = 'normalized'
+        grp.attrs['training_time'] = time.strftime("%Y-%m-%d %H:%M:%S", time.gmtime())
+        grp.attrs['final_training_loss'] = history.history['loss'][-1]
+        grp.attrs['final_validation_loss'] = history.history['val_loss'][-1]
```
Comment (LiveMethod) [inline train.py:225]: Not doing anything with these, but copilot suggested them and they seemed neat, lol

File: train.py [hunk 16]
```diff
@@ -1,216 +1,228 @@
...
     model.add(LSTM(hidden_units))
     model.add(Dense(1, activation=None))
-    model.compile(optimizer=Adam(learning_rate=learning_rate), loss=error_to_signal, metrics=[error_to_signal])
-    print(model.summary())
-
-    # Load and Preprocess Data ###########################################
-    in_rate, in_data = wavfile.read(args.in_file)
-    out_rate, out_data = wavfile.read(args.out_file)
+    model.compile(optimizer=Adam(learning_rate=learning_rate), loss=Data.esr, metrics=[Data.esr])
+    print('\n') ;model.summary(); print('\n')
```
Comment (LiveMethod) [inline train.py:101]: model.summary() does it's own printing, so this double prints and appends a stray "None" after the summary which was confusing. Dropped and added new line on either side.
